Added new lint: `reserve_after_initialization`

Red-Rapious opened the pull request:
Closes https://github.com/rust-lang/rust-clippy/issues/11330.

A new lint that informs the user about a more concise way to create a vector with a known capacity.
Example:
```rust
let mut v: Vec<usize> = vec![];
v.reserve(10);
```

Produces the following help:
```rust
  |
2 | /     let mut v: Vec<usize> = vec![];
3 | |     v.reserve(10);
  | |__________________^ help: consider using `Vec::with_capacity(space_hint)`: `let v: Vec<usize> = Vec::with_capacity(10);`
  |
```

And can be rewritten as:
```rust
let v: Vec<usize> = Vec::with_capacity(10);
```

changelog: new lint [`reserve_after_initialization`]

y21 commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,225 @@
+            let adjusted_mut = adjusted_ty.ref_mutability().unwrap_or(Mutability::Not);
+            needs_mut |= adjusted_mut == Mutability::Mut;
+            match parent.kind {
+                ExprKind::AddrOf(_, Mutability::Mut, _) => {
+                    needs_mut = true;
+                    return ControlFlow::Break(true);
+                },
+                ExprKind::Unary(UnOp::Deref, _) | ExprKind::Index(..) if !needs_mut => {
+                    let mut last_place = parent;
+                    while let Some(parent) = get_parent_expr(cx, last_place) {
+                        if matches!(parent.kind, ExprKind::Unary(UnOp::Deref, _) | ExprKind::Field(..))
+                            || matches!(parent.kind, ExprKind::Index(e, _, _) if e.hir_id == last_place.hir_id)
+                        {
+                            last_place = parent;
+                        } else {
+                            break;
+                        }
+                    }
+                    needs_mut |= cx.typeck_results().expr_ty_adjusted(last_place).ref_mutability()
+                        == Some(Mutability::Mut)
+                        || get_parent_expr(cx, last_place)
+                            .map_or(false, |e| matches!(e.kind, ExprKind::AddrOf(_, Mutability::Mut, _)));
+                },
+                ExprKind::MethodCall(_, recv, ..)
+                    if recv.hir_id == e.hir_id
+                        && adjusted_mut == Mutability::Mut
+                        && !adjusted_ty.peel_refs().is_slice() =>
+                {
+                    // No need to set `needs_mut` to true. The receiver will be either explicitly borrowed, or it will
+                    // be implicitly borrowed via an adjustment. Both of these cases are already handled by this point.
+                    return ControlFlow::Break(true);
+                },
+                ExprKind::Assign(lhs, ..) if e.hir_id == lhs.hir_id => {
+                    needs_mut = true;
+                    return ControlFlow::Break(false);
+                },
+                _ => (),
+            }
+            ControlFlow::Continue(())
+        });
```
I think this doesn't handle  `let ref mut y = v;` correctly (i.e. will not detect `v` as requiring mutability after a reserve call).
But also, this mutability checking seems like a lot of code with lots of logic for a suggestion that isn't machine applicable. I feel like that's not really necessary for a `HasPlaceholders` suggestion and clippy could just leave `mut` out and expect the user to fill it in if needed, or maybe `snippet` could be used to get the source string from `let` to `=`.

Red-Rapious replied:
Thank you for your feedback! I will indeed remove this `mut` checking and let the user adapt it.

Red-Rapious replied:
I decided to go for the `snippet` option instead, and added 3 more tests to show the different cases.

y21 reviewed (commented):
Looks pretty good to me already for a first contribution 👀 but I have one question

blyxyas commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,139 @@
+    /// Informs the user about a more concise way to create a vector with a known capacity.
+    ///
+    /// ### Why is this bad?
+    /// The `Vec::with_capacity` constructor is easier to understand.
```
"easier to understand" seems kinda subjective, asserting that it's `less complex` or `more readable` would transmit the idea better.

y21 commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,139 @@
+            self.searcher = Some(VecReserveSearcher {
+                local_id: id,
+                err_span: local.span,
+                init_part: format!("let {}: {}", 
+                snippet(cx, local.pat.span, ""),
+                match local.ty {
+                    Some(type_inference) => snippet(cx, type_inference.span, "").to_string(),
+                    None => String::new()
+                }),
```
Instead of getting two snippets here, this could probably pass a single larger span to `snippet`, from `let` to the low part of the initializer expression. This might simplify it a bit further and also make it slightly nicer in case there is no type annotation (right now it would suggest `let x: = Vec::with_capacity()`)
```suggestion
                init_part: snippet(cx, local.span.shrink_to_lo().to(init_expr.span.shrink_to_lo()), "").into_owned(),
```
This is what `init_part` would store with this change:
```rs
let mut v = Vec::new();
^^^^^^^^^^^^
```
Then, in `display_err` it could be just `format!("{}Vec::with_capacity({})", self.init_part, self.space_hint)` (without the `=` sign and whitespaces, since that will already be there).
This would also need to be done in `check_expr` if this is implemented in this place.

(It might even be possible to use `write!` or `String::push_str` instead of `format!` since this is just appending to an already existing string)

y21 replied:
Oh, the init_expr span part probably needs to be `init_expr.span.source_callsite().shrink_to_lo()`, since the initializer expression can be part of a macro expansion (`vec![]`). In those cases we need the real source span that invoked that macro and not some span inside the `vec![]` macro.

blyxyas commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,139 @@
+            && let PatKind::Binding(BindingAnnotation::MUT, id, _name, None) = local.pat.kind
+            && !in_external_macro(cx.sess(), local.span)
+            && let Some(init) = get_vec_init_kind(cx, init_expr)
+            && !matches!(init, VecInitKind::WithExprCapacity(_))
+            && !matches!(init, VecInitKind::WithConstCapacity(_))
```
This can be optimized with some magic patterns

```suggestion
            && !matches!(init, VecInitKind::WithExprCapacity(_)
               | VecInitKind::WithConstCapacity(_)
            )
```

blyxyas replied:
This still needs to be addressed. Having some pattern magic here (`|`) makes faster both compiling (less macros) and more runtime (less branches)

Red-Rapious replied:
My bad, I changed it in `check_expr` but not in `check_local`.

blyxyas commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,139 @@
+
+    fn check_local(&mut self, cx: &LateContext<'tcx>, local: &'tcx Local<'tcx>) {
+        if let Some(init_expr) = local.init
+            && let PatKind::Binding(BindingAnnotation::MUT, id, _name, None) = local.pat.kind
```
This `_name` is never used, it can be discarded.
```suggestion
            && let PatKind::Binding(BindingAnnotation::MUT, id, _, None) = local.pat.kind
```

blyxyas commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,139 @@
+                local_id: id,
+                err_span: local.span,
+                init_part: format!("let {}: {}", 
+                snippet(cx, local.pat.span, ""),
+                match local.ty {
+                    Some(type_inference) => snippet(cx, type_inference.span, "").to_string(),
```
AFAIK it's good practice to use `".."` as snippet defaults. If that default should never be met, you can check for the snippet with [`snippet_opt`](https://doc.rust-lang.org/beta/nightly-rustc/clippy_utils/source/fn.snippet_opt.html) in the initial_matching.

blyxyas commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,139 @@
+            RESERVE_AFTER_INITIALIZATION,
+            self.err_span,
+            "calls to `reverse` immediately after creation",
+            "consider using `Vec::with_capacity(space_hint)`",
```
`rust-analyzer` in IDE integrations can automatically place the cursor over the placeholder if you use this format. (If the user has integrated Clippy with their RA)

```suggestion
            "consider using `Vec::with_capacity(/* Space hint */)`",
```

blyxyas commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,139 @@
+            cx,
+            RESERVE_AFTER_INITIALIZATION,
+            self.err_span,
+            "calls to `reverse` immediately after creation",
```
```suggestion
            "call to `reserve` immediately after creation",
```

blyxyas commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,139 @@
+        if self.searcher.is_none()
+            && let ExprKind::Assign(left, right, _) = expr.kind
+            && let ExprKind::Path(QPath::Resolved(None, path)) = left.kind
+            && let [_name] = &path.segments
```
This named pattern destructuring could just be a wildcard, as it isn't used elsewhere. It will still have the same destructuring meaning.

```suggestion
            && let [_] = &path.segments
```

blyxyas commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,139 @@
+                && let ExprKind::MethodCall(name, self_arg, other_args, _) = expr.kind
+                && other_args.len() == 1
+                && path_to_local_id(self_arg, searcher.local_id)
+                && name.ident.as_str() == "reserve"
```
Using the [`sym!`](https://doc.rust-lang.org/beta/nightly-rustc/clippy_utils/macro.sym.html) macro here would be faster (`Symbol::as_str` is slow as it has to lock the symbol interner). Creating a symbol from a string is faster than creating a string from a symbol :cat:

```suggestion
                && name.ident.as_str() == sym!(reserve)
```

y21 replied:
Hmm, doesn't `sym!` do the same (locking the interner)? 🤔 
https://doc.rust-lang.org/beta/nightly-rustc/src/rustc_span/symbol.rs.html#1914-1916

blyxyas replied:
Mmmm... You're right, `sym!` expands to `intern` that also uses `self.0.lock()`. Maybe it should be mentioned on the `sym!` docs. / `intern`'s documentation. So yeah, it doesn't matter if it's changed.

blyxyas commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,139 @@
+            {
+                self.searcher = Some(VecReserveSearcher {
+                    err_span: searcher.err_span.to(stmt.span),
+                    space_hint: snippet(cx, other_args[0].span, "").to_string(),
```
```suggestion
                    space_hint: snippet(cx, other_args[0].span, "..").to_string(),
```

y21 commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,139 @@
+    fn check_stmt(&mut self, cx: &LateContext<'tcx>, stmt: &'tcx Stmt<'_>) {
+        if let Some(searcher) = self.searcher.take() {
+            if let StmtKind::Expr(expr) | StmtKind::Semi(expr) = stmt.kind
+                && let ExprKind::MethodCall(name, self_arg, other_args, _) = expr.kind
+                && other_args.len() == 1
```
mini nit
```suggestion
                && let ExprKind::MethodCall(name, self_arg, [space_hint], _) = expr.kind
```
then `space_hint` can be used instead of `other_args[0]`

blyxyas reviewed (commented):
WOW, really cool for a first lint, just some very minor nits, but very **very** good overall :heart:, I'm really impressed!

blyxyas commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,139 @@
+}
+impl VecReserveSearcher {
+    fn display_err(&self, cx: &LateContext<'_>) {
+        if self.space_hint.is_empty() {
```
```suggestion
        if self.space_hint.is_empty() || is_from_proc_macro(err_span) {
```

blyxyas commented on tests/ui/reserve_after_initialization.rs:
This should include some tests with macros, you can get some inspiration with other test files like `tests/ui/allow_attributes.rs` (that's a simple one)

Red-Rapious replied:
How do I manually test an example file containing a macro? When running:
```console
$ cargo dev lint tests/ui/allow_attributes.rs
```
I get the following error:
```rust
error[E0463]: can't find crate for `proc_macros`
 --> tests/ui/allow_attributes.rs:7:1
  |
7 | extern crate proc_macros;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate
```

blyxyas replied:
To run a specific lint use the [`TESTNAME` env. var.](https://doc.rust-lang.org/cargo/commands/cargo-test.html) and the `cargo uitest` command.

```sh
$ TESTNAME="allow_attributes" cargo uitest
```

Using `cargo dev lint` is used to execute your lint on a normal file AFAIK, it isn't really used (at least I've never used it)

y21 commented on tests/ui/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,19 @@
+#![warn(clippy::reserve_after_initialization)]
```
A test for assign expressions would be nice to have.  Something like
```rs
let mut x: Vec<usize> = Vec::new();
x = Vec::new();
x.reserve(10);
```

Red-Rapious commented:
I added some tests and fixed everything you mentioned except for:
- the use of `syml!`, since it apparently will not improve performance (I'm glad, I couldn't get it to work properly anyway)
- the macros tests: I have trouble using an external crate while running tests manually, as I explained above

y21 commented:
Btw I think you can mark things that are done as resolved, that might make it a little easier to read through and see what's left ^^ (saying this because my very first comment is taking up lots of space and that comment is not relevant anymore/already applied, I can't click "resolve" myself it looks like)

blyxyas commented:
I just noticed that actually testing with macros isn't really documented anywhere, so I'll do a tutorial here.

In our test suite headers can be used with the "//@" syntax. E.g. "//@run-rustfix" as the first line is a header.
You can use the header `//@aux-build:proc_macros.rs:proc-macro` to import the auxiliary create just for testing procedural macros and how lints interact with them.

So, after using `//@aux-build:proc_macros.rs:proc-macro` you can then `extern macro proc_macro;`, and then `use proc_macros::{external, with_span}`.

The auxiliary proc macro `external!` simulates a proc macro that is external to the crate (I think), and the `with_span` macro simulates a proc macro that changes the span, so you can verify if you're linting the correct expression and such.

We really should document this, #10597 
If in the future something isn't in the book or other documentation (and a reviewer isn't online), you can always check the `rustc-dev-guide`, it's a book mainly for developing rustc and maybe cargo, but it has some things that also apply to Clippy. (if a reviewer is online, you can always ask for help here or on the [Zulip server](https://rust-lang.zulipchat.com/), we all check that chat more or less periodically :heart: )

y21 commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,139 @@
+            && let PatKind::Binding(BindingAnnotation::MUT, id, _, None) = local.pat.kind
+            && !in_external_macro(cx.sess(), local.span)
+            && let Some(init) = get_vec_init_kind(cx, init_expr)
+            && !matches!(init, VecInitKind::WithExprCapacity(_)
+               | VecInitKind::WithConstCapacity(_)
+            )
```
```suggestion
            && !matches!(init, VecInitKind::WithExprCapacity(_) | VecInitKind::WithConstCapacity(_))
```
Sadly rustfmt doesn't know about let chains yet so this formatting needs to be done manually for now😢

y21 commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,139 @@
+        if self.searcher.is_none()
+            && let ExprKind::Assign(left, right, _) = expr.kind
+            && let ExprKind::Path(QPath::Resolved(None, path)) = left.kind
+            && let [_] = &path.segments
```
Is this `let [_]` needed? Can there ever be more or less than one segment in a path to a local or what does this do?

Red-Rapious replied:
I previously used this `let` to retrieve the name of the first segment, but I think there's not use to check that `path.segments.len() == 1`; I'll remove the line.

y21 commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,139 @@
+            {
+                self.searcher = Some(VecReserveSearcher {
+                    err_span: searcher.err_span.to(stmt.span),
+                    space_hint: snippet(cx, space_hint.span, "..").to_string(),
```
```suggestion
                    space_hint: snippet(cx, space_hint.span, "..").into_owned(),
```
This allows it to reuse the string if owned (which afaik is basically always the case) and is consistent with the other usages of snippet here

y21 reviewed (commented):
Just the last few tiny nits I could find ^^. After these it looks good to me 👍

blyxyas commented on clippy_lints/src/reserve_after_initialization.rs:
```diff
@@ -0,0 +1,134 @@
+            if let StmtKind::Expr(expr) | StmtKind::Semi(expr) = stmt.kind
+                && let ExprKind::MethodCall(name, self_arg, [space_hint], _) = expr.kind
+                && path_to_local_id(self_arg, searcher.local_id)
+                && !is_from_proc_macro(cx, expr)
```
`is_from_proc_macro` should go last, as it's very expensive, there are some efforts on optimizing it, but for the moment, it should go last.

blyxyas reviewed (approved):
Just this mini-nit and it should be good to go! Really impressive for a first contribution :heart:

y21 reviewed (approved):
LGTM!

> your reviews were very helpful!

Thanks :D Glad to be of help

blyxyas commented:
@bors r=blyxyas,y21

bors commented:
:hourglass: Testing commit f3c58773026bf042142be2bf8ed1e8b8797a4f68 with merge 4932d0573360bb71d078579842d7b7712edff1a3...
<!-- homu: {"type":"BuildStarted","head_sha":"f3c58773026bf042142be2bf8ed1e8b8797a4f68","merge_sha":"4932d0573360bb71d078579842d7b7712edff1a3"} -->

Red-Rapious commented:
Thank you so much for your help, @blyxyas and @y21!
